### ISSUE TITLE
O_NERD and O_EMOJI creates duplicate letters in filename

### ISSUE THREAD

**Issue by FelixKratz** [jarun/nnn#1692]
#### Environment details (Put `x` in the checkbox along with the information)

- [x] Operating System: macOS Ventura
- [x] Desktop Environment: Aqua
- [x] Terminal Emulator: kitty, Alacritty
- [x] Shell: zsh
- [ ] Custom desktop opener (if applicable):
- [x] Program options used: none
- [x] Configuration options set: O_NERD
- [ ] Plugins are installed
- [x] Issue exists on `nnn` master

#### Exact steps to reproduce the issue
Clone the repo, enable nerdfonts in the `makefile`
```makefile
O_NERD := 1  # support icons-nerdfont
```
and compile with `LDLIBS="-L/opt/homebrew/opt/ncurses/lib/" make`.

Some files (with certain icons it appears) have duplicate initial letters as seen in the screenshot below:
![Screenshot 2023-07-14 at 11 48 58](https://github.com/jarun/nnn/assets/22680421/a01ccdcf-55b0-4a7e-bfb2-a0152226fc9e)

as soon as I select the makefile it will also have the duplicate initial letter:
![Screenshot 2023-07-14 at 11 54 34](https://github.com/jarun/nnn/assets/22680421/9801c5e9-2fe1-4a66-b458-145b471e69ba)

I have done some testing, this seems to also happen with `O_EMOJI` is set to enabled:
![Screenshot 2023-07-14 at 11 56 51](https://github.com/jarun/nnn/assets/22680421/ce931411-2488-4bfb-ba9a-46b0bd2f7396)

 but not with `O_ICONS` enabled:
![Screenshot 2023-07-14 at 11 57 37](https://github.com/jarun/nnn/assets/22680421/765c2c14-f100-4e8d-b1e3-6e65be515474)

I have peeked into the source code to find what is different between those options, but did not find a convincing clue.

I have also checked with kitty and alacritty, they both face the same problem.

These are the dynamic libs used by nnn:
```bash
 x otool -L ./nnn
./nnn:
	/usr/lib/libedit.3.dylib (compatibility version 2.0.0, current version 3.0.0)
	/opt/homebrew/opt/ncurses/lib/libncursesw.6.dylib (compatibility version 6.0.0, current version 6.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```

**Comment by N-R-K** [comment]
Previous discussion: https://github.com/jarun/nnn/discussions/1662

According to https://github.com/jarun/nnn/discussions/1662#discussioncomment-6147447 v4.7 didn't have this problem (can you confirm this too? @FelixKratz) but I went through the diff between 4.7 and 4.8 and didn't notice anything that might cause it.

Also another person who reported the issue was on macos as well (https://github.com/jarun/nnn/discussions/1662#discussioncomment-6298138) so maybe something in macos is causing it?

**Comment by FelixKratz** [comment]
Sorry, I did not search the discussion, only the issues.

I have done a quick bisect and this is the commit where it goes wrong: https://github.com/jarun/nnn/commit/20e944f5e597239ed491c213a634eef3d5be735e

**Comment by FelixKratz** [comment]
Interestingly, `O_EMOJI` shows this problem since the introduction in https://github.com/jarun/nnn/commit/b45be9bc5b34c863db8c68e56fedc4112d582191.

**Comment by jarun** [comment]
I use emojis on xfce4-terminal and don't see this issue.
Can you try to fix this and raise a PR? I will try on my setup and see if it works for me as well?
The change should be in `print_icon()` and/or `adjust_cols()`. See if changing the padding works.

**Comment by FelixKratz** [comment]
I believe this problem has no trivial fix.

It seems that removing
```c
	if (sel)
		attrs |= A_REVERSE;
```
in `printent` fixes the duplication problem. This suggests to me that there is a problem with the ncurses internal redraw and damage logic. This is possibly connected to a disagreement in the terminals perceived width for the character and the ncurses calculated width of the character. But to be honest, I have no clue about terminal rendering.

**Comment by jarun** [comment]
@luukvbaal would you able to check this?

**Comment by jarun** [comment]
@FelixKratz what happens if you replace `A_REVERSE` with `A_STANDOUT`?

**Comment by FelixKratz** [comment]
> @FelixKratz what happens if you replace `A_REVERSE` with `A_STANDOUT`?

As soon as any attrs are set that would cause the filename to be redrawn it starts to happen

No attrs:
<img width="412" alt="Screenshot 2023-07-16 at 15 49 44" src="https://github.com/jarun/nnn/assets/22680421/e8f0acbb-2ab3-49ab-84ac-182ed8b3c58e">

`A_STANDOUT`:
<img width="412" alt="Screenshot 2023-07-16 at 15 50 34" src="https://github.com/jarun/nnn/assets/22680421/7675b58e-cdea-44b9-8125-966d66c26a67">

`A_BOLD`:
<img width="412" alt="Screenshot 2023-07-16 at 15 50 55" src="https://github.com/jarun/nnn/assets/22680421/d94b7830-3735-462a-afef-0051162c0e3d">

**Comment by jarun** [comment]
@N-R-K @luukvbaal what's the best course of action here?

We can't make the next release without fixing this.

I see something here that I don't fully understand: https://github.com/jarun/nnn/discussions/1661

**Comment by N-R-K** [comment]
> All instances where this is a problem wcswidth reports a length of 2 as seen in the screenshot below, where I display the wcswidth of the icon to the left of the actual item: ![Screenshot 2023-07-14 at 15 08 29](https://user-images.githubusercontent.com/22680421/253580356-b269bcde-17fc-4e01-9228-164edbac1072.png)

Interesting. This *does* seem to be what's causing the issue, because on my system (linux/glibc), all those icons have a width of 1 when called in `wcswidth`.

![image](https://github.com/jarun/nnn/assets/79544946/1fb8c0c1-ebb6-427e-87a4-2d856aba5b07)

Could this be a libc bug in macos? Can you share the output of `locale`? Also what do you see when you run `LC_ALL="C.UTF8" nnn`?

**Comment by FelixKratz** [comment]
> Could this be a libc bug in macos? Can you share the output of `locale`? Also what do you see when you run `LC_ALL="C.UTF8" nnn`?

My `locale` output is:
<img width="232" alt="Screenshot 2023-07-16 at 21 46 06" src="https://github.com/jarun/nnn/assets/22680421/aa0f30f1-5ae3-46ec-894c-55cf6b57d8ef">

Running `LC_ALL="C.UTF8" ./nnn` looks completely wild:
<img width="440" alt="Screenshot 2023-07-16 at 21 51 43" src="https://github.com/jarun/nnn/assets/22680421/c87083e5-d125-4735-86a9-768bc7cf0dc8">

I will happily try out any ideas, I am a bit lost regarding this problem as I have no experience with ncurses.

**Comment by lawnowner** [comment]
I don't think your problem is due to a bug in nnn. Unless you can reproduce this on Linux, maybe your OS is using a different text normalization for unicode encoding (e.g. NFD instead of NFC), so the width of some characters are 2 instead of 1 for you due to decomposition.

The locale category values in double quotes are not set explicitly, they are implied values derived from LANG or LC_ALL, but in your output all values are in double quotes, so none of them are explicitly set in your environment. You can try `LC_CTYPE="en_US.UTF-8" ./nnn` or `LC_ALL="en_US.UTF-8" ./nnn` on the command line, or setting the `LC_` variables in your shell startup file and/or terminal emulator config. If you got a weird `utf-8-mac` variant of the locale, you can try that as well.

**Comment by cyaconi** [comment]
I just tried both `LC_CTYPE="en_US.UTF-8" ./nnn` and `LC_ALL="en_US.UTF-8" ./nnn` in macOS and the problem continues. Hope it helps.

**Comment by FelixKratz** [comment]
I seem to have found a fix for this: #1694

**Comment by quantonganh** [comment]
I confirmed that `wcswidth` returns the different value for the same Unicode code point between macOS and Linux.

macOS:

```
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

<img width="220" alt="Screen Shot 2023-08-07 at 11 25 54" src="https://github.com/jarun/nnn/assets/1568504/5e7e0e02-9ad7-4782-aff3-381829def2a2">

--- 

Linux:

```
LANG=en_GB.UTF-8
LANGUAGE=
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```

<img width="215" alt="Screen Shot 2023-08-07 at 11 26 23" src="https://github.com/jarun/nnn/assets/1568504/8d23b36e-92b4-4b36-815a-aef4d4bd69cc">

---

In version 4.8, the code point is different, width is 1 but the icon is missing:

<img width="214" alt="Screen Shot 2023-08-07 at 11 47 57" src="https://github.com/jarun/nnn/assets/1568504/11db455f-102e-4beb-bea9-3963d102d813">

---

> I don't think your problem is due to a bug in nnn. Unless you can reproduce this on Linux, maybe your OS is using a different text normalization for unicode encoding (e.g. NFD instead of NFC), so the width of some characters are 2 instead of 1 for you due to decomposition.

It could be the result of a combination of `nnn` and the Nerd fonts, locale settings, unicode version and the C library implementation.
